indent/ecrystal.vim

```diff
@@ -0,0 +1,487 @@
+" Setup {{{1
+" =====
+
+if exists('b:did_indent')
+  finish
+endif
+
+call ecrystal#SetSubtype()
+
+if b:ecrystal_subtype != ''
+  exec 'runtime! indent/'.b:ecrystal_subtype.'.vim'
+  unlet! b:did_indent
+endif
+
+if &l:indentexpr == ''
+  if &l:cindent
+    let &l:indentexpr = 'cindent(v:lnum)'
+  else
+    let &l:indentexpr = 'indent(prevnonblank(v:lnum - 1))'
+  endif
+endif
+
+let b:ecrystal_subtype_indentexpr = &l:indentexpr
+
+" Should we use folding?
+if has('folding') && get(g:, 'ecrystal_fold', 0)
+  setlocal foldmethod=expr
+  setlocal foldexpr=GetEcrystalFold()
+endif
+
+" Should closing control tags be aligned with their corresponding
+" opening tags?
+if !exists('b:ecrystal_align_end')
+  if exists('g:ecrystal_align_end')
+    let b:ecrystal_align_end = g:ecrystal_align_end
+  else
+    let b:ecrystal_align_end = b:ecrystal_subtype !=# 'html' && b:ecrystal_subtype !=# 'xml'
+  endif
+endif
+
+" Should multiline tags be indented?
+if !exists('b:ecrystal_indent_multiline')
+  let b:ecrystal_indent_multiline = get(g:, 'ecrystal_indent_multiline', 1)
+endif
+
+if b:ecrystal_indent_multiline
+  runtime! indent/crystal.vim
+  unlet! b:did_indent
+  setlocal indentexpr<
+endif
+
+setlocal indentexpr=GetEcrystalIndent()
+setlocal indentkeys+=<>>,=end,=else,=elsif,=rescue,=ensure,=when
+
+let b:did_indent = 1
+
+" Only define the function once.
+if exists('*GetEcrystalIndent')
+  finish
+endif
+
+" Helper variables and functions {{{1
+" ==============================
+
+const s:ecr_open = '<%%\@!'
+const s:ecr_close = '%>'
+
+const s:ecr_control_open = '<%%\@!-\=[=#]\@!'
+const s:ecr_comment_open = '<%%\@!-\=#'
+
+const s:ecr_indent_regex =
+      \ '\<\%(if\|unless\|else\|elsif\|case\|for\|when\|while\|until\|begin\|do\|rescue\|ensure\|' .
+      \ 'class\|module\|struct\|lib\|enum\|union\)\>'
+
+const s:ecr_dedent_regex =
+      \ '\<\%(end\|else\|elsif\|when\|rescue\|ensure\)\>'
+
+" Return the value of a single shift-width
+if exists('*shiftwidth')
+  let s:sw = function('shiftwidth')
+else
+  function s:sw()
+    return &shiftwidth
+  endfunction
+endif
+
+" Return the current cursor position
+function s:Cursor()
+  return [line('.'), col('.')]
+endfunction
```

Comment:
Would you use `getcurpos()` persistently for getting cursor position rather than making our own utility?